src/net/server/middleware/tsig.rs

```diff
@@ -0,0 +1,506 @@
+            Ok(Some(Self::mk_signed_truncated_response(
+                request,
+                truncation_ctx,
+            )?))
+        } else {
+            Ok(None)
+        }
+    }
+
+    fn mk_signed_truncated_response(
+        request: &Request<RequestOctets>,
+        truncation_ctx: TruncationContext<KS::Key, tsig::Key>,
+    ) -> Result<AdditionalBuilder<StreamTarget<NextSvc::Target>>, ServiceError>
+    {
+        let builder = mk_builder_for_target();
+        let mut new_response = builder
+            .start_answer(request.message(), Rcode::NOERROR)
+            .unwrap();
+        new_response.header_mut().set_tc(true);
+        let mut additional = new_response.additional();
+
+        match truncation_ctx {
+            TruncationContext::HaveSigner(signer) => {
+                if let Err(err) =
+                    signer.answer(&mut additional, Time48::now())
+                {
+                    error!("Unable to sign truncated TSIG response: {err}");
+                    Err(ServiceError::InternalError)
+                } else {
+                    Ok(additional)
+                }
+            }
+
+            TruncationContext::NoSignerOnlyTheKey(key) => {
+                // We can't use the TSIG signer state we just had as that was
+                // consumed in the failed attempt to sign the answer, so we
+                // have to create a new TSIG state in order to sign the
+                // truncated response.
+                let octets = request.message().as_slice().to_vec();
+                let mut mut_msg = Message::from_octets(octets).unwrap();
+
+                match ServerTransaction::request(
+                    &key,
+                    &mut mut_msg,
+                    Time48::now(),
+                ) {
+                    Ok(None) => {
+                        error!("Unable to create signer for truncated TSIG response: internal error: request is not signed but was expected to be");
+                        Err(ServiceError::InternalError)
+                    }
+
+                    Err(err) => {
+                        error!("Unable to create signer for truncated TSIG response: {err}");
+                        Err(ServiceError::InternalError)
+                    }
+
+                    Ok(Some(signer)) => {
+                        if let Err(err) =
+                            signer.answer(&mut additional, Time48::now())
+                        {
+                            error!("Unable to sign truncated TSIG response: {err}");
+                            Err(ServiceError::InternalError)
+                        } else {
+                            Ok(additional)
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    fn map_stream_item(
+        request: Request<RequestOctets, ()>,
+        stream_item: ServiceResult<NextSvc::Target>,
+        pp_config: &mut PostprocessingConfig<KS::Key>,
```

Comment:
Not sure how much of this is based on previous convention, but a mutable config is a bit weird to me. It also has some strange implications for other bits where theres now `&mut ()` and `&mut bool`. I wonder if there's some other way to implement that kind of caching behaviour that you're doing here.

Reply:
It was just the fall out of realizing that `Arc<Mutex<..>>` wasn't need to ensure mutability and that mutability is only there to allow an `Option` member in a struct to be `take()`n. It's not about caching (or at least not what I think of as caching, not sure what you had in mind when using that term). This has changed several times already, and there's some talk of maybe revisiting the streaming side of the `Service` interface in the future to split out the common single request -> response case from the single request -> multiple/streaming response case, and this post-processing map stuff is all about the latter case. So long answer short, I don't have an immediate idea for how to do this differently and this stuff is all behind an "unstable" feature gate as it's a design in-progress so we may well come back and change this.

Reply:
Maybe caching was too ambiguous, I meant the "store result for future invocations" part, which is also why this is mutable right?

Reply:
It was called config as it was originally configuration to drive the post-processing phase, but it became mutable because that config for TSIG includes a signer which may need replacing with a different one if the response is a stream not a single response. There was some discussion about maybe reworking the underlying TSIG module code to present a single signing interface for single and multi-response but that is out of scope for now. If that already existed the mutation here wouldn't happen. For now I can rename config  to state as that's what it effectively became instead of the original intent.

Reply:
If it's gonna change later I'm not too worried about it at all. Thanks for the response!

Reply:
Actually, I'm getting mixed up. That is a thing (replacing the signer) but in this case it's just because the signer needs to be consumed so taken out of the struct, but I can't consume the entire struct.

Reply:
Maybe it should really have two parts: immutable config, and mutable state. But I don't really want to get into refining this further right now, I think it's better to let this evolve as we add more middleware that needs to post-process, or we revisit this entirely as we go as mentioned.